crates/rsonpath-test-codegen/src/compression.rs

```diff
@@ -0,0 +1,228 @@
+//! Compressing JSON files while maintaining integrity of expected results.
+//! 
+//! We first compress by mapping each byte with information on whether it is significant
+//! or should be removed. Using that annotation we can figure out the expected byte offsets
+//! of results in the compressed version with respect to the original.
+use crate::{
+    files::Files,
+    model::{self},
+    DiscoveredDocument,
+};
+use std::{io, string::FromUtf8Error};
+
+/// Read all existing TOML documents and generate the compressed variants.
+pub(crate) fn generate_compressed_documents(files: &mut Files) -> Result<(), io::Error> {
+    let original_documents: Vec<_> = files.documents().into_iter().cloned().collect();
+
+    original_documents
+        .into_iter()
+        .filter(|doc| !doc.document.input.is_compressed)
+        .try_for_each(|doc| compress_document(files, &doc))
+}
+
+fn compress_document(files: &mut Files, doc: &DiscoveredDocument) -> Result<(), io::Error> {
+    let (required_file, contents) = match &doc.document.input.source {
+        model::InputSource::LargeFile(f) => (Some(f), files.read_json(f)?),
+        model::InputSource::JsonString(c) => (None, c.clone()),
+    };
+
+    let compressed_input = CompressedInput::new(&contents);
+    let queries = doc
+        .document
+        .queries
+        .iter()
+        .map(|q| {
+            let compressed_results = compressed_input.transform_results(&q.results);
+
+            model::Query {
+                results: compressed_results,
+                description: q.description.clone(),
+                query: q.query.clone(),
+            }
+        })
+        .collect();
+
+    let json_string = compressed_input.into_string();
+    let source = if let Some(f) = required_file {
+        let relative_path = files.add_compressed_large_json(f, json_string);
+        model::InputSource::LargeFile(relative_path)
+    } else {
+        model::InputSource::JsonString(json_string)
+    };
+
+    let compressed_doc = model::Document {
+        input: model::Input {
+            description: format!("{} (compressed)", doc.document.input.description),
+            is_compressed: true,
+            source,
+        },
+        queries,
+    };
+
+    files.add_compressed_document(&doc.relative_path, doc.name.clone(), compressed_doc);
+
+    Ok(())
+}
+
+/// Compressed JSON input.
+struct CompressedInput {
+    json_string: JsonString,
+}
+
+/// Categorization of JSON bytes into whitespace and significant bytes.
+#[derive(Clone, Copy)]
+enum JsonByte {
+    /// The byte is significant and should be preserved.
+    Significant(u8),
+    /// The byte is some whitespace and does not matter.
+    Whitespace,
+}
+
+/// Wrapper over a vec of bytes to implement traits.
+struct JsonString(Vec<JsonByte>);
+
+impl From<JsonString> for Vec<u8> {
+    fn from(value: JsonString) -> Self {
+        value.0.into_iter().filter_map(|x| x.byte()).collect()
```

Comment:
Highly subjective with no real benefit, but worth considering the stylistic aspects of using `byte` directly.
https://rust-lang.github.io/rust-clippy/master/index.html#/redundant_closure

Wdyt?

Reply:
This one is a different lint, actually: https://rust-lang.github.io/rust-clippy/master/index.html#/redundant_closure_for_method_calls

More importantly, you've reminded me that I didn't enable my pedantic lints that are otherwise enabled in the lib 😄 So I enabled them and fixed violations, which includes this particular case.